optimum/habana/transformers/generation/utils.py

```diff
@@ -170,6 +170,50 @@ def _update_model_kwargs_for_generation(
 
         return model_kwargs
 
+    def _prepare_decoder_attention_mask(
+        self,
+        max_steps: int,  #current stopping criteria
+        batch_size: int,
+        pad_token_id: int,
+        device: str,
+        dtype: str = torch.int64,
+    ) -> torch.Tensor:
+        x = torch.zeros((batch_size, max_steps), device=device, dtype=dtype)
+        return x.index_fill(1, torch.tensor([0]), pad_token_id)    # First the position with pad_token_id
+
+    def _prepare_past_key_values(
+        self,
+        max_steps: int,  #current stopping criteria
+        batch_size: int,
+        device: str,
+    ) -> Tuple[torch.LongTensor, torch.LongTensor, Optional[torch.Tensor], Optional[torch.Tensor]]:
+        n_heads = self.config.decoder_attention_heads
+        embed_size_per_head = self.config.d_model // n_heads
+        key_states_ = torch.zeros((self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device)
+        value_states_ = torch.zeros((self.config.decoder_layers, batch_size, n_heads, max_steps, embed_size_per_head), device=device)
+        past_key_values = tuple((key_states_[i, :, :, :, :], value_states_[i, :, :, :, :], None, None) for i in range(self.config.decoder_layers))
+        return past_key_values
```

Comment:
Past key values should have the right shape since `input_ids` and `attention_mask` are padded no?

Reply:
No, As far as I understand. Past_key_values are concatened for every token generation. Hence, this is needed for us to get specifical value from index without change in shape